design-proposals/klip-0-add-protobuf-support.md

```diff
@@ -0,0 +1,77 @@
+# KLIP 0 - Add Protobuf SerDe Support
+
+Author: [Stephen Powis](http://github.com/crim)
+
+Release target: 5.1
+
+Status: In Discussion 
+
+Discussion: _link to the design discussion PR_
+
+## tl;dr
+
+Add support to KSQL to process messages stored using the [ProtocolBuffers](https://developers.google.com/protocol-buffers/) serialization format.   
+
+## Motivation and background
+
+[ProtocolBuffers](https://developers.google.com/protocol-buffers/) is a language and platform neutral serialization format that is commonly used with Apache Kafka. Adding out-of-the-box support for this message format would expand the environments in which KSQL can be utilized.
+
+## What is in scope
+
+Adding SerDe support for ProtocolBuffers Version 3 to KSQL.
+
+## What is not in scope
+
+Supporting streams/topics containing multiple ProtocolBuffer schema types.  _See [Design](#Design) section below for more details._
+
+## Value/Return
+
+Out of the box support for running KSQL against Apache Kafka topics containing messages encoded with ProtocolBuffers.
+
+## Public APIS
+
+Proposed extending the `CREATE STREAM` grammar to include a new _value_format_ type **PROTOBUF** as well as adding a new _tableProperty_ named **protobuf_class**.
+
+Example:
+```genericsql
+ksql> CREATE STREAM stream_name (....) WITH (kafka_topic='TopicName', value_format='PROTOBUF', protobuf_class='org.whatever.protobufs.MyProtoBuf$EncodedMessageType');
```

Comment:
Would the fields available on the `EncodedMessageType` define the schema of the stream? i.e. would this work like Avro currently does, in that the user does not need to supply the list of columns and their types when working with Avro?

Reply:
Also, what if the class is not in the class path? How does the user know which classes are available? I think thinking through the UX here is worth while. We may choose to implement that UX in a staged manner, but we should have a shared agreement of where we want to get to here.

Reply:
> Would the fields available on the EncodedMessageType define the schema of the stream?

Yes with all Protobuf semantics in place.

> would this work like Avro currently does, in that the user does not need to supply the list of columns and their types when working with Avro?

Yes.

Reply:
@apurvam Any comments?